File: frontend/src/lib/api/merge-fields.ts
/**
 * Merge Fields API Client
 */

import axios from "axios";
import type {
  MergeField,
  MergeFieldCreate,
  MergeFieldUpdate,
  MergeFieldListResponse,
  MergeFieldDiscoveryResult,
  MergeFieldDataType,
} from "@/types/v2";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

const api = axios.create({
  baseURL: `${API_BASE_URL}/api`,
});

export interface MergeFieldListParams {
  category?: string;
  search?: string;
  data_type?: MergeFieldDataType;
  is_iterable?: boolean;
  page?: number;
  per_page?: number;
}

export const mergeFieldsApi = {
  /**
   * List merge fields with optional filters
   */
  async list(params?: MergeFieldListParams): Promise<MergeFieldListResponse> {
    const { data } = await api.get<MergeFieldListResponse>("/merge-fields", {
      params,
    });
    return data;
  },

  /**
   * Get a single merge field by ID
   */
  async get(fieldId: string): Promise<MergeField> {
    const { data } = await api.get<MergeField>(`/merge-fields/${fieldId}`);
    return data;
  },

  /**
   * Create a new merge field
   */
  async create(payload: MergeFieldCreate): Promise<MergeField> {
    const { data } = await api.post<MergeField>("/merge-fields", payload);
    return data;
  },

  /**
   * Update a merge field
   */
  async update(fieldId: string, payload: MergeFieldUpdate): Promise<MergeField> {
    const { data } = await api.put<MergeField>(`/merge-fields/${fieldId}`, payload);
    return data;
  },

  /**
   * Delete a merge field
   */
  async delete(fieldId: string): Promise<void> {
    await api.delete(`/merge-fields/${fieldId}`);
  },

  /**
   * Get list of categories
   */
  async getCategories(): Promise<string[]> {
    const { data } = await api.get<{ categories: string[] }>("/merge-fields/categories");
    return data.categories;
  },

  /**
   * Search for autocomplete suggestions
   */
  async autocomplete(query: string, limit: number = 10): Promise<MergeField[]> {
    const { data } = await api.get<MergeField[]>("/merge-fields/autocomplete", {
      params: { q: query, limit },
    });
    return data;
  },

  /**
   * Trigger merge field discovery scan
   */
  async scan(createMissing: boolean = true): Promise<MergeFieldDiscoveryResult> {
    const { data } = await api.post<MergeFieldDiscoveryResult>("/merge-fields/scan", null, {
      params: { create_missing: createMissing },
    });
    return data;
  },
};

File: backend/app/routers/merge_fields.py
"""
Merge Fields API Routes
"""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.merge_field import (
    MergeFieldCreate,
    MergeFieldDiscoveryResult,
    MergeFieldListResponse,
    MergeFieldResponse,
    MergeFieldUpdate,
)
from app.services.merge_field_service import MergeFieldService

router = APIRouter(prefix="/merge-fields", tags=["Merge Fields"])


@router.get("", response_model=MergeFieldListResponse)
async def list_merge_fields(
    db: AsyncSession = Depends(get_db),
    category: str | None = Query(None),
    search: str | None = Query(None),
    data_type: str | None = Query(None),
    is_iterable: bool | None = Query(None),
    page: int = Query(1, ge=1),
    per_page: int = Query(50, ge=1, le=100),
):
    """List all merge fields with optional filters."""
    fields, total = await MergeFieldService.get_list(
        db, category=category, search=search, data_type=data_type, is_iterable=is_iterable, page=page, per_page=per_page
    )

    return MergeFieldListResponse(
        merge_fields=[MergeFieldResponse.model_validate(f) for f in fields],
        total=total,
        page=page,
        per_page=per_page,
        total_pages=(total + per_page - 1) // per_page if total > 0 else 0,
    )


@router.get("/categories")
async def get_categories(db: AsyncSession = Depends(get_db)):
    """Get list of all unique categories."""
    categories = await MergeFieldService.get_categories(db)
    return {"categories": categories}


@router.get("/autocomplete")
async def autocomplete(
    db: AsyncSession = Depends(get_db), q: str = Query(..., min_length=1), limit: int = Query(10, ge=1, le=50)
):
    """Search merge fields for autocomplete suggestions."""
    fields = await MergeFieldService.search_autocomplete(db, q, limit)
    return [MergeFieldResponse.model_validate(f) for f in fields]


@router.post("/scan", response_model=MergeFieldDiscoveryResult)
async def trigger_discovery(db: AsyncSession = Depends(get_db), create_missing: bool = Query(True)):
    """Scan all templates and discover merge fields."""
    result = await MergeFieldService.discover_all(db, create_missing=create_missing)
    return MergeFieldDiscoveryResult(**result)


@router.get("/{field_id}", response_model=MergeFieldResponse)
async def get_merge_field(field_id: UUID, db: AsyncSession = Depends(get_db)):
    """Get a merge field by ID."""
    field = await MergeFieldService.get_by_id(db, field_id)
    if not field:
        raise HTTPException(status_code=404, detail="Merge field not found")
    return MergeFieldResponse.model_validate(field)


@router.post("", response_model=MergeFieldResponse, status_code=201)
async def create_merge_field(body: MergeFieldCreate, db: AsyncSession = Depends(get_db)):
    """Create a new merge field."""
    existing = await MergeFieldService.get_by_path(db, body.path)
    if existing:
        raise HTTPException(status_code=409, detail=f"Path '{body.path}' already exists")

    field = await MergeFieldService.create(db, **body.model_dump())
    return MergeFieldResponse.model_validate(field)


@router.put("/{field_id}", response_model=MergeFieldResponse)
async def update_merge_field(field_id: UUID, body: MergeFieldUpdate, db: AsyncSession = Depends(get_db)):
    """Update a merge field."""
    field = await MergeFieldService.get_by_id(db, field_id)
    if not field:
        raise HTTPException(status_code=404, detail="Merge field not found")

    field = await MergeFieldService.update(db, field, **body.model_dump(exclude_unset=True))
    return MergeFieldResponse.model_validate(field)


@router.delete("/{field_id}", status_code=204)
async def delete_merge_field(field_id: UUID, db: AsyncSession = Depends(get_db)):
    """Delete a merge field."""
    field = await MergeFieldService.get_by_id(db, field_id)
    if not field:
        raise HTTPException(status_code=404, detail="Merge field not found")

    await MergeFieldService.delete(db, field)
    return None
